icicle-compiler/test/Icicle/Test/Arbitrary/Program.hs

```diff
@@ -4,201 +4,594 @@
 import qualified Icicle.Avalanche.Program as A
 import qualified Icicle.Avalanche.Simp as A
 import qualified Icicle.Avalanche.Statement.Flatten as A
+import qualified Icicle.Core.Eval.Program   as Core
+import qualified Icicle.Core.Exp.Prim      as Core
+import qualified Icicle.Core.Program.Check as Core
+import qualified Icicle.Core.Program.Program as Core
+import qualified Icicle.Sea.Eval as Sea
+import qualified Icicle.Sea.FromAvalanche.State as Sea
+import qualified Icicle.Sea.Data as Sea
+import qualified Icicle.Compiler as Compiler
+
+import           Control.Monad.IO.Class (liftIO)
+import           Control.Arrow ((&&&))
+
+import           Data.Map (Map)
+import           Data.Maybe
+import           Data.String
+import qualified Data.List as List
+import qualified Data.Map as Map
+import qualified Data.Text as Text
+import qualified Data.Text.Lazy as LT
+import qualified Data.List.NonEmpty as NonEmpty
 
-import           Icicle.Common.Base
-import           Icicle.Common.Eval
-import           Icicle.Common.Type
-import           Icicle.Common.Annot
+import           Text.Show.Pretty (ppShow)
 
-import           Icicle.Internal.Pretty
+import           P
 
-import qualified Icicle.Sea.FromAvalanche.Analysis as S
-import qualified Icicle.Sea.Eval as S
+import qualified Prelude as Savage
 
-import           Icicle.Test.Arbitrary.Data
-import           Icicle.Test.Arbitrary.Core
+import           System.Directory (getTemporaryDirectory, removeDirectoryRecursive)
+import           System.IO
+import           System.IO.Temp (createTempDirectory)
 
-import           P
+import           Disorder.Corpus
 
-import           Test.QuickCheck
 
-import qualified Prelude as Savage
+import           Test.QuickCheck (Gen, Arbitrary(..), elements, getPositive, discard)
+import           Test.QuickCheck (listOf, listOf1)
 
-import           Text.Show.Pretty (ppShow)
+import           X.Control.Monad.Trans.Either (EitherT, bracketEitherT')
 
 
-newtype InputType = InputType {
-    unInputType :: ValType
+-- | inputType = SumT ErrorT factType
+newtype SumErrorFactT = SumErrorFactT {
+    factType :: ValType
   } deriving (Show)
 
 data WellTyped = WellTyped {
-    wtEntities      :: [Entity]
-  , wtInputId       :: InputId
-  , wtFactType      :: ValType
-  , wtFacts         :: [AsAt BaseValue]
-  , wtTime          :: Time
-  , wtMaxMapSize    :: Int
-  , wtCore          :: C.Program ()         Var
-  , wtAvalanche     :: A.Program ()         Var C.Prim
-  , wtAvalancheFlat :: A.Program (Annot ()) Var A.Prim
+    wtFacts    :: [WellTypedValue]
+  , wtClusters :: [WellTypedCluster]
+  }
+
+-- FIXME
+--
+-- This only gnerates one compute kernel for each cluster.
+-- PsvFission tests multiple kernels by doing its own thing.
+-- Perhaps roll that into this as well.
+--
+data WellTypedCluster = WellTypedCluster {
+    wtCluster       :: Sea.Cluster () ()
+  , wtFactType      :: !ValType -- clusterInputType == SumT ErrorT wtFactType
```

Comment:
nah, it's the value type `t` as it would be coming from a TOML dictionary. A cluster input type is `SumT ErrorT t` and a core program stream type is `SumT ErrorT (t, TimeT`)` (or something like that)